### ISSUE TITLE
Error when cloning on Windows with Git 2.24: filename in tree entry contains backslash -- solved in Git 2.25

### ISSUE THREAD

**Issue by carlescufi** [civetweb/civetweb#812]
The project triggers an error when cloning on Windows when using Git 2.24 or higher:
`filename in tree entry contains backslash: '\'`

- Potential offending commit in civetweb: https://github.com/civetweb/civetweb/commit/4ac467e3e07a2abb02b894c0b508e9885ee4d7d8
- Offending commit in Git 2.24: https://github.com/git/git/commit/e1d911dd4c7b76a5a8cec0f5c8de15981e34da83
- Git for Windows issue: https://github.com/git-for-windows/git/issues/2435

**Workaround**

Disable [core.protectNTFS](https://git-scm.com/docs/git-config#Documentation/git-config.txt-coreprotectNTFS):
`git config --global core.protectNTFS false`

Additional details can be found here:
https://github.com/zephyrproject-rtos/zephyr/issues/21426

**Comment by bel2125** [comment]
I checked the file [这是一个示例文本.html](https://github.com/civetweb/civetweb/blob/master/test/nonlatin/%E8%BF%99%E6%98%AF%E4%B8%80%E4%B8%AA%E7%A4%BA%E4%BE%8B%E6%96%87%E6%9C%AC.html) in the commit https://github.com/civetweb/civetweb/commit/4ac467e3e07a2abb02b894c0b508e9885ee4d7d8. 
The file name is "this is a sample text" in Chinese characters, but all characters are in the [BMP](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane), so we do not even need [Surrogates](https://docs.microsoft.com/en-us/windows/win32/intl/surrogates-and-supplementary-characters) here ... **and there is definitely no backslash** in the file name. CivetWeb has this file in a test folder, to check if non-ascii URLs work like expected.
The file and this test are perfectly valid.

I have an older version of git on Windows and it works there ... so, it seems to me, this is a recent bug in git (for Windows), not a bug in CivetWeb.

**Comment by carlescufi** [comment]
@bel2125 thanks for the reply.
This is not clearly a bug in Git for Windows, since it addresses a vulnerability. See the discussion in [this issue](https://github.com/git-for-windows/git/issues/2435) for more information.
The issue in the civetweb repository might not be related to the file in commit https://github.com/civetweb/civetweb/commit/4ac467e3e07a2abb02b894c0b508e9885ee4d7d8, but perhaps somewhere else in the repo's history, since this will be checked for all tree objects. There is almost certainly some point in the repo's history where a file with backslashes was added, but we haven't been able to find which exactly.

**Comment by carlescufi** [comment]
@bel2125 @mbolivar
could the offending file be the one that was removed by 83dc425e256667de837334a608598ca7b348fe92?

**Comment by bel2125** [comment]
This `test/\/a.txt` dates back to 2012, years before CivetWeb was forked from Mongoose.
Today CivetWeb has 586 public forks on GitHub, several private forks on GitHub and probably countless copies in other places (e.g. in Git repositories hosted locally in companies).
I think "rewrite history" is way off the scale.

Do you have a reference to this Git issue?

**Comment by bel2125** [comment]
> Do you have a reference to this Git issue?

You already posted it ... https://github.com/git-for-windows/git/issues/2435

I wonder if CivetWeb is the only repository with this kind of issue.

**Comment by carlescufi** [comment]
> I wonder if CivetWeb is the only repository with this kind of issue.

Not quite, no, though it doesn't seem to be very common:
https://stackoverflow.com/questions/59345571/pycharm-behaving-strange-with-git-filename-in-tree-entry-contains-backslash
https://github.com/confluentinc/cp-ansible/issues/167

**Comment by bel2125** [comment]
https://github.com/git-for-windows/git/issues/2435 is probably fixed in Git for Windows 2.25 (fixed in v2.25.0-rc1.windows.1)

#809 seems to be a different issue - it does not use Git for Windows and problem seems to be *.cmd files there.

It's a Git issue, not a CivetWeb issue. Should we keep this issue open until Git for Windows 2.25 is released? Should we add some comment in the README not to use V2.24 with CivetWeb?

**Comment by mbolivar** [comment]
> Should we keep this issue open until Git for Windows 2.25 is released? Should we add some comment in the README not to use V2.24 with CivetWeb?

IMO, that would be a reasonable way to resolve this.

**Comment by bel2125** [comment]
Do you know what versions of Git for Windows are affected? V2.24 ... what about earlier versions V2.23, V2.22, ...? 
Do you know when V2.25 will be available (approximately)?

**Comment by mbolivar** [comment]
> Do you know what versions of Git for Windows are affected? V2.24 ... what about earlier versions V2.23, V2.22, ...?

I believe it's just 2.24.

> Do you know when V2.25 will be available (approximately)?

https://github.com/git-for-windows/git/issues/2435#issuecomment-570717462

**Comment by carlescufi** [comment]
@bel2125 I agree with the proposal. A brief mention that Git 2.24 has an issue with Civetweb on Windows should cover it.

**Comment by bel2125** [comment]
Done:
https://github.com/civetweb/civetweb/commit/61a868be17596f124f3c2737e0e06a6ad6f25cd1

Let's keep this note for a couple of months, ... until we can assume most of the developers updated.

**Comment by neutrinog** [comment]
Hey guys. I don't actually use civetweb or really know what it is, but I stumbled upon this thread while researching this same issue. This does appear to be fixed in git 2.25 and you can find notes about it in the [2.25 Release Notes](https://raw.githubusercontent.com/git/git/master/Documentation/RelNotes/2.25.0.txt):

```
 * An earlier update to Git for Windows declared that a tree object is
   invalid if it has a path component with backslash in it, which was
   overly strict, which has been corrected.  The only protection the
   Windows users need is to prevent such path (or any path that their
   filesystem cannot check out) from entering the index.
   (merge 224c7d70fa js/mingw-loosen-overstrict-tree-entry-checks later to maint).
```
 Cheers!

**Comment by bel2125** [comment]
Solved in Git 2.25
A note not to use Git 2.24 will remain in the README for a while.
